File: test/ModularPipelines.UnitTests/ReturnNothingTests.cs
using ModularPipelines.Models;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests;

public class ReturnNothingTests : TestBase
{
    private class ReturnNothingModule1 : SimpleTestModule<CommandResult>
    {
        protected override CommandResult? Result => null;
    }

    private class ReturnNothingModule2 : SimpleTestModule<CommandResult>
    {
        protected override CommandResult? Result => null;
    }

    private class ReturnNothingModule3 : SimpleTestModule<CommandResult>
    {
        protected override CommandResult? Result => default;
    }

    [Test]
    public async Task Module1_HasValue_False()
    {
        var result = await await RunModule<ReturnNothingModule1>();

        await Assert(result);
    }

    [Test]
    public async Task Module2_HasValue_False()
    {
        var result = await await RunModule<ReturnNothingModule2>();

        await Assert(result);
    }

    [Test]
    public async Task Module3_HasValue_False()
    {
        var result = await await RunModule<ReturnNothingModule3>();

        await Assert(result);
    }

    private static async Task Assert(ModuleResult<CommandResult?> result)
    {
        using (TUnit.Assertions.Assert.Multiple())
        {
            await TUnit.Assertions.Assert.That(result.IsSuccess).IsTrue();
            await TUnit.Assertions.Assert.That(result.ModuleResultType).IsEqualTo(ModuleResultType.Success);
            await TUnit.Assertions.Assert.That(result.ValueOrDefault).IsNull();
            await TUnit.Assertions.Assert.That(result.ExceptionOrDefault).IsNull();
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Results/ReturnNothingTests.fs
namespace ModularPipelines.UnitTests.FSharp.Results

open ModularPipelines.Models
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type private ReturnNothingModule1() =
    inherit SimpleTestModule<CommandResult>()
    override _.Result = Unchecked.defaultof<CommandResult>

type private ReturnNothingModule2() =
    inherit SimpleTestModule<CommandResult>()
    override _.Result = Unchecked.defaultof<CommandResult>

type private ReturnNothingModule3() =
    inherit SimpleTestModule<CommandResult>()
    override _.Result = Unchecked.defaultof<CommandResult>

type ReturnNothingTests() =
    inherit TestBase()

    member private _.AssertResult(result: ModuleResult<CommandResult>) = async {
        do! check(Assert.That(result.IsSuccess).IsTrue())
        do! check(Assert.That(result.ModuleResultType = ModuleResultType.Success).IsTrue())
        do! check(Assert.That(result.ValueOrDefault).IsNull())
        do! check(Assert.That(result.ExceptionOrDefault).IsNull())
    }

    [<Test>]
    member this.Module1_HasValue_False() = async {
        let! m = this.RunModule<ReturnNothingModule1>() |> Async.AwaitTask
        let! result = m.CompletionSource.Task |> Async.AwaitTask
        do! this.AssertResult(result)
    }

    [<Test>]
    member this.Module2_HasValue_False() = async {
        let! m = this.RunModule<ReturnNothingModule2>() |> Async.AwaitTask
        let! result = m.CompletionSource.Task |> Async.AwaitTask
        do! this.AssertResult(result)
    }

    [<Test>]
    member this.Module3_HasValue_False() = async {
        let! m = this.RunModule<ReturnNothingModule3>() |> Async.AwaitTask
        let! result = m.CompletionSource.Task |> Async.AwaitTask
        do! this.AssertResult(result)
    }
